csharp/src/Apache.Arrow.Flight/Middleware/Interceptors/ClientInterceptorAdapter.cs

```diff
@@ -0,0 +1,161 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Apache.Arrow.Flight.Middleware.Interfaces;
+using Grpc.Core;
+using Grpc.Core.Interceptors;
+
+namespace Apache.Arrow.Flight.Middleware.Interceptors
+{
+    public sealed class ClientInterceptorAdapter : Interceptor
+    {
+        private readonly IReadOnlyList<IFlightClientMiddlewareFactory> _factories;
+
+        public ClientInterceptorAdapter(IEnumerable<IFlightClientMiddlewareFactory> factories)
+        {
+            _factories = factories?.ToList() ?? throw new ArgumentNullException(nameof(factories));
+        }
+
+        public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
+            TRequest request,
+            ClientInterceptorContext<TRequest, TResponse> context,
+            AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
+            where TRequest : class
+            where TResponse : class
+        {
+            var options = InterceptCall(context, out var middlewares);
+
+            var newContext = new ClientInterceptorContext<TRequest, TResponse>(
+                context.Method,
+                context.Host,
+                options);
+
+            var call = continuation(request, newContext);
+
+            return new AsyncUnaryCall<TResponse>(
+                HandleResponse(call.ResponseAsync, call.ResponseHeadersAsync, call.GetStatus, call.GetTrailers,
+                    call.Dispose, middlewares),
+                call.ResponseHeadersAsync,
+                call.GetStatus,
+                call.GetTrailers,
+                call.Dispose
+            );
+        }
+
+        public override AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRequest, TResponse>(
+            TRequest request,
+            ClientInterceptorContext<TRequest, TResponse> context,
+            AsyncServerStreamingCallContinuation<TRequest, TResponse> continuation)
+            where TRequest : class
+            where TResponse : class
+        {
+            var callOptions = InterceptCall(context, out var middlewares);
+            var newContext = new ClientInterceptorContext<TRequest, TResponse>(
+                context.Method, context.Host, callOptions);
+
+            var call = continuation(request, newContext);
+
+            var responseHeadersTask = call.ResponseHeadersAsync.ContinueWith(task =>
+            {
+                if (task.Exception == null && task.Result != null)
+                {
+                    var headers = task.Result;
+                    foreach (var m in middlewares)
+                        m?.OnHeadersReceived(new CallHeaders(headers));
+                }
+
+                return task.Result;
```

Comment:
If `task.Exception` can be not-null, then should this throw the exception instead of returning null?